src/main/scala/org/constellation/domain/redownload/MissingProposalFinder.scala

```diff
@@ -0,0 +1,98 @@
+package org.constellation.domain.redownload
+
+import cats.syntax.all._
+import org.constellation.domain.redownload.RedownloadService.{PeersCache, PeersProposals, SnapshotsAtHeight}
+import org.constellation.schema.Id
+import org.constellation.schema.snapshot.{HeightRange, MajorityInfo}
+
+import scala.collection.immutable.NumericRange.inclusive
+import scala.math.{max, min}
+
+class MissingProposalFinder(
+  private val heightInterval: Long,
+  private val offset: Long,
+  private val limit: Long
+) {
+
+  /**
+    * @return Map where the key is the Id of a peer and the value is a set of heights with missing proposals from the peer
+    */
+  def findMissingPeerProposals(
+    majorityRange: HeightRange,
+    proposals: PeersProposals,
+    peersCache: PeersCache
+  ): Map[Id, Set[Long]] = {
+    val receivedProposals = proposals.mapValues(_.keySet)
+    val bounds = boundedRange(majorityRange)
+
+    peersCache
+      .mapValues(
+        _.map { majorityHeight =>
+          HeightRange(majorityHeight.joined.getOrElse(0L), majorityHeight.left.getOrElse(Long.MaxValue))
```

Comment:
You have already introduced `HeightRange.MaxValue` so you can reuse it here with help of `Applicative[Option]`.

```scala
Applicative[Option]
     .map2(majorityHeight.joined, majorityHeight.left)(HeightRange)
     .getOrElse(HeightRange.MaxValue)
```

or even cleaner with `cats.syntax`:

```scala
import cats.syntax._

(majorityHeight.joined, majorityHeight.left)
   .mapN(HeightRange)
   .getOrElse(HeightRange.MaxValue)
```

Thanks to that you can keep default values of `HeightRange` in one place

Reply:
True, I should reuse the `MaxValue` here.

Reply:
Actually, the question is if we can have `left` without `joined` :P 

Currently, it assumes that both can be None independently :D